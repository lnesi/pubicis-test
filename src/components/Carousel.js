import React, { useReducer, useEffect } from 'react';
import { ReactComponent as ArrowSVG } from '../assets/arrow.svg';
import './Carousel.scss';

export const Image = ({ webformatURL, tags }) => {
  return (
    <div className="App__Carousel__Images__Item">
      <div
        className="App__Carousel__Images__Item__Image"
        style={{ backgroundImage: `url(${webformatURL})` }}
      >
        <img src={webformatURL} alt={tags} data-testid="carousel-item-image" />
      </div>
      <h3>{tags}</h3>
    </div>
  );
};

export function paginationReducer(state, action) {
  let currentPage = state.currentPage;
  switch (action.type) {
    case 'RESIZE':
      let screen = getScreen(action.payload);
      return {
        ...state,
        screen: screen,
        currentPage: 0,
        page: getPage({ ...state, screen, currentPage: 0 }),
        screenWidth: action.payload
      };
    case 'NEXT':
      currentPage = state.currentPage + 1;
      return {
        ...state,
        currentPage,
        page: getPage({ ...state, currentPage })
      };
    case 'PREV':
      currentPage = state.currentPage - 1;
      return {
        ...state,
        currentPage,
        page: getPage({ ...state, currentPage })
      };
    default:
      return state;
  }
}

export function getPage(state) {
  return state.images.slice(
    state.currentPage * PAGE_SIZES[state.screen],
    state.currentPage * PAGE_SIZES[state.screen] + PAGE_SIZES[state.screen]
  );
}

export function getScreen(width) {
  let screen = 'mobile';
  if (width >= 768 && width < 1024) {
    screen = 'tablet';
  } else if (width >= 1024 && width < 1200) {
    screen = 'desktop';
  } else if (width >= 1200 && width < 1440) {
    screen = 'xl';
  } else if (width >= 1440) {
    screen = 'xxl';
  }
  return screen;
}

export function initPagination(state) {
  return { ...state, page: getPage(state) };
}

export const PAGE_SIZES = {
  mobile: 1,
  tablet: 2,
  desktop: 3,
  xl: 4,
  xxl: 5
};

export default function Carousel({ images }) {
  const [state, dispatch] = useReducer(
    paginationReducer,
    { images, page: [], screen: 'mobile', currentPage: 0 },
    initPagination
  );

  useEffect(() => {
    const onResize = event => {
      dispatch({ type: 'RESIZE', payload: event.target.innerWidth });
    };
    if (window) {
      window.addEventListener('resize', onResize);
      dispatch({ type: 'RESIZE', payload: window.innerWidth });
    }
    return () => {
      window.removeEventListener('resize', onResize);
    };
  }, [dispatch]);

  return (
    <div className="App__Carousel" data-testid="Carousel">
      <div className="App__Carousel__Images">
        {state.page.map((image, i) => {
          return (
            <Image
              key={`image_${i}_${image.id}`}
              {...image}
              width={100 / PAGE_SIZES[state.screen]}
            />
          );
        })}
      </div>
      <div className="App__Carousel__Controls">
        <button
          className="App__Carousel__Controls__Prev"
          disabled={!(state.currentPage > 0)}
          onClick={() => {
            dispatch({ type: 'PREV' });
          }}
        >
          <span>Prev</span>
          <ArrowSVG />
        </button>
        <button
          className="App__Carousel__Controls__Next"
          disabled={
            !(
              state.currentPage <
              Math.ceil(state.images.length / PAGE_SIZES[state.screen]) - 1
            )
          }
          onClick={() => {
            dispatch({ type: 'NEXT' });
          }}
        >
          <span>Next</span>
          <ArrowSVG />
        </button>
      </div>
    </div>
  );
}
