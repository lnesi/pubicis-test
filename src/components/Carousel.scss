@import '../scss/variables';

.App__Carousel {
  &__Images {
    display: flex;
    position: absolute;
    width: 100%;
    box-sizing: border-box;
    @media only screen and (min-width: $mobile-breakpoint) {
      padding: 2rem;
      position: relative;
    }

    &__Item {
      color: $fg-color;
      width: 100%;
      h3 {
        margin-left: 1rem;
        @media only screen and (min-width: $mobile-breakpoint) {
          margin-left: 0;
        }
      }
      &__Image {
        background-repeat: no-repeat;
        background-size: cover;
        background-position: 50%;
      }
      img {
        width: 100%;
        opacity: 0;
        height: 50vh;
        @media only screen and (min-width: $mobile-breakpoint) {
          height: 250px;
        }
      }
      @media only screen and (min-width: $mobile-breakpoint) {
        margin-right: 0.5rem;
        margin-left: 0.5rem;
      }
    }
  }

  &__Controls {
    box-sizing: border-box;
    position: absolute;
    width: 100%;
    height: 400px;
    overflow: hidden;
    @media only screen and (min-width: $mobile-breakpoint) {
      height: auto;
      text-align: center;
      background-color: $bg-color;
      padding: 1rem;
    }
    button {
      font-family: Times, 'Times New Roman', serif;
      font-weight: bold;
      background-color: rgba(255, 255, 255, 0.5);
      color: white;
      padding: 0.5rem 1rem;
      border: none;
      height: 100px;
      width: 100px;
      font-size: 1rem;
      cursor: pointer;
      position: absolute;
      margin-top: 25%;
      display: inline-block;
      @media only screen and (min-width: $mobile-breakpoint) {
        position: relative;
        background-color: $fg-color;
        margin-top: 0;
        height: auto;
        svg {
          display: none;
        }
      }
      svg {
        width: 30px;
        fill: $fg-color;
      }
      &:disabled {
        opacity: 0.3;
      }
    }
    span {
      display: none;
      @media only screen and (min-width: $mobile-breakpoint) {
        display: inline;
      }
    }
    &__Prev {
      margin-left: -50px;
      border-top-right-radius: 50%;
      border-bottom-right-radius: 50%;
      svg {
        transform: rotate(180deg);
        margin-left: 33px;
      }

      @media only screen and (min-width: $mobile-breakpoint) {
        border-radius: unset;
        border-top-left-radius: 5px;
        border-bottom-left-radius: 5px;
        margin-right: 0.25rem;
      }
    }
    &__Next {
      right: 0;
      margin-right: -50px;
      border-top-left-radius: 50%;
      border-bottom-left-radius: 50%;
      svg {
        margin-right: 33px;
      }
      @media only screen and (min-width: $mobile-breakpoint) {
        border-radius: unset;
        border-top-right-radius: 5px;
        border-bottom-right-radius: 5px;
        margin-left: 0.25rem;
      }
    }
  }
}
