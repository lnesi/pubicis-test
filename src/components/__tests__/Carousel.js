import React from 'react';
import { render } from '@testing-library/react';

import Carousel, {
  Image,
  getPage,
  getScreen,
  initPagination,
  paginationReducer,
  PAGE_SIZES
} from '../Carousel';
const screenTestCases = {
  480: 'mobile',
  600: 'mobile',
  780: 'tablet',
  1000: 'tablet',
  1100: 'desktop',
  1250: 'xl',
  1600: 'xxl'
};
describe('Helper Functions', () => {
  Object.keys(PAGE_SIZES).forEach(size => {
    it(`getPage returns correct for screen ${size}`, () => {
      const state = {
        currentPage: 2,
        images: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15],
        screen: size
      };
      const page = getPage(state);
      expect(page.length).toBe(PAGE_SIZES[size]);
    });
  });
  Object.keys(screenTestCases).forEach(size => {
    it('getScreen returns correct size for: ' + size, () => {
      const screen = getScreen(size);
      expect(screen).toBe(screenTestCases[size]);
    });
  });

  it('initPagination returns mobile first', () => {
    const state = {
      currentPage: 2,
      images: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15],
      screen: 'mobile'
    };
    const initializedState = initPagination(state);
    expect(initializedState.page).toStrictEqual(getPage(state));
  });
});

describe('Reducer', () => {
  it('Handles Action NEXT', () => {
    const state = {
      images: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15],
      screen: 'tablet',
      currentPage: 0
    };
    const newState = paginationReducer(state, { type: 'NEXT' });
    expect(newState.currentPage).toBe(1);
    expect(newState.page).toStrictEqual([2, 3]);
  });

  it('Handles Action PREV', () => {
    const state = {
      images: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15],
      screen: 'desktop',
      currentPage: 3
    };
    const newState = paginationReducer(state, { type: 'PREV' });
    expect(newState.currentPage).toBe(2);
    expect(newState.page).toStrictEqual([6, 7, 8]);
  });

  it('Handles Action RESIZE', () => {
    const state = {
      images: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15],
      screen: 'mobile',
      currentPage: 3
    };
    const newState = paginationReducer(state, {
      type: 'RESIZE',
      payload: 1500
    });
    expect(newState.currentPage).toBe(0);
    expect(newState.page).toStrictEqual([0, 1, 2, 3, 4]);
    expect(newState.screen).toBe('xxl');
  });
  it('Handles Default action', () => {
    const state = {
      images: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15],
      screen: 'mobile',
      currentPage: 3
    };
    const newState = paginationReducer(state, {
      type: 'UNKOWN'
    });
    expect(newState).toStrictEqual(state);
  });
});

describe('render', () => {
  it('Image component renders', () => {
    const { getByText, getByTestId } = render(
      <Image
        webformatURL="http://www.test.com/image.jpg"
        tags="test tag text"
      />
    );
    const linkElement = getByText(/test tag text/i);
    const imgElement = getByTestId('carousel-item-image');
    expect(linkElement).toBeInTheDocument();
    expect(imgElement).toBeInTheDocument();
  });

  it('Carousel componsnt renders', () => {
    const images = [];
    for (let i = 0; i < 20; i++) {
      images.push({
        id: i + 1,
        webformatURL: `http://www.test.com/image_${i + 1}.jpg`,
        tags: 'test tag ' + (i + 1)
      });
    }
    const { getByTestId } = render(<Carousel images={images} />);
    const carousel = getByTestId('Carousel');
    expect(carousel).toBeInTheDocument();
  });
});

describe('snapshots', () => {
  it('Image component match snapshot', () => {
    const { container } = render(
      <Image
        webformatURL="http://www.test.com/image.jpg"
        tags="test tag text"
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('Carousel componsnt match snapshot', () => {
    const images = [];
    for (let i = 0; i < 20; i++) {
      images.push({
        id: i + 1,
        webformatURL: `http://www.test.com/image_${i + 1}.jpg`,
        tags: 'test tag ' + (i + 1)
      });
    }
    const { container } = render(<Carousel images={images} />);
    expect(container).toMatchSnapshot();
  });
});
