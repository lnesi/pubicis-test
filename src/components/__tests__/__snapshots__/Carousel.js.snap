// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots Carousel componsnt match snapshot 1`] = `
<div>
  <div
    class="App__Carousel"
    data-testid="Carousel"
  >
    <div
      class="App__Carousel__Images"
    >
      <div
        class="App__Carousel__Images__Item"
      >
        <div
          class="App__Carousel__Images__Item__Image"
          style="background-image: url(http://www.test.com/image_1.jpg);"
        >
          <img
            alt="test tag 1"
            data-testid="carousel-item-image"
            src="http://www.test.com/image_1.jpg"
          />
        </div>
        <h3>
          test tag 1
        </h3>
      </div>
      <div
        class="App__Carousel__Images__Item"
      >
        <div
          class="App__Carousel__Images__Item__Image"
          style="background-image: url(http://www.test.com/image_2.jpg);"
        >
          <img
            alt="test tag 2"
            data-testid="carousel-item-image"
            src="http://www.test.com/image_2.jpg"
          />
        </div>
        <h3>
          test tag 2
        </h3>
      </div>
      <div
        class="App__Carousel__Images__Item"
      >
        <div
          class="App__Carousel__Images__Item__Image"
          style="background-image: url(http://www.test.com/image_3.jpg);"
        >
          <img
            alt="test tag 3"
            data-testid="carousel-item-image"
            src="http://www.test.com/image_3.jpg"
          />
        </div>
        <h3>
          test tag 3
        </h3>
      </div>
    </div>
    <div
      class="App__Carousel__Controls"
    >
      <button
        class="App__Carousel__Controls__Prev"
        disabled=""
      >
        <span>
          Prev
        </span>
        <svg>
          arrow.svg
        </svg>
      </button>
      <button
        class="App__Carousel__Controls__Next"
      >
        <span>
          Next
        </span>
        <svg>
          arrow.svg
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`snapshots Image component match snapshot 1`] = `
<div>
  <div
    class="App__Carousel__Images__Item"
  >
    <div
      class="App__Carousel__Images__Item__Image"
      style="background-image: url(http://www.test.com/image.jpg);"
    >
      <img
        alt="test tag text"
        data-testid="carousel-item-image"
        src="http://www.test.com/image.jpg"
      />
    </div>
    <h3>
      test tag text
    </h3>
  </div>
</div>
`;
